src/component/mxgraph/shape/render/BpmnCanvas.ts

```diff
@@ -42,23 +42,32 @@ export interface BpmnCanvasConfiguration {
  */
 export default class BpmnCanvas {
   private c: mxgraph.mxXmlCanvas2D; // TODO use mxAbstractCanvas2D when fully available in mxgraph-type-definitions
+
+  private readonly iconOriginalSize: Size;
   private readonly scaleX: number;
   private readonly scaleY: number;
 
   private iconPaintingOriginX = 0;
   private iconPaintingOriginY = 0;
 
-  private shapeConfiguration: ShapeConfiguration;
+  private readonly shapeConfiguration: ShapeConfiguration;
 
   constructor(config: BpmnCanvasConfiguration) {
     this.c = config.mxCanvas;
     this.shapeConfiguration = config.shapeConfiguration; // TODO clone?
 
     const parentSize = Math.min(this.shapeConfiguration.w, this.shapeConfiguration.h);
     const iconConfiguration = config.iconConfiguration;
-    const iconOriginalSize = iconConfiguration.originalSize;
-    this.scaleX = (parentSize / iconOriginalSize.width) * iconConfiguration.ratioFromShape;
-    this.scaleY = (parentSize / iconOriginalSize.height) * iconConfiguration.ratioFromShape;
+    this.iconOriginalSize = iconConfiguration.originalSize;
+
+    const ratioFromShape = iconConfiguration.ratioFromShape;
+    if (ratioFromShape) {
+      this.scaleX = (parentSize / this.iconOriginalSize.width) * ratioFromShape;
+      this.scaleY = (parentSize / this.iconOriginalSize.height) * ratioFromShape;
+    } else {
+      this.scaleX = 1;
```

Comment:
I don't know if the ratio from parent is calculated like I do in **IconPainter.calculateIconSize(...)**.
But when the _ratioFromParent_ parameter is equals to **1**, I consider that the icon must have the same size as the shape (if height < width of shape, icon height = shape height and icon width is calculate proportionally with original icon height / shape height).

I don't know if it's really clear. We can refactor the scale calculation in another PR 😉

Reply:
@csouchet I plan to do "no more actual icon size computation in IconPainter" in the next refactoring as described in #247
The computation is not done in the same way but should. This will be part of the refactoring or a next one to unify this.
The idea here is to do the refactoring without behaviour changes with baby steps. So for now, there are potentially duplications, several ways of doing things (like before). This will be managed progressively.


About your remark about the ratioFromParent = 1 (remember this is only temporar to make the icon painter uses the BPMN Canvas but the integration is not finished)
- the actual icon size is computed outside of the canvas, using the ratioFromParent
- the icon size passed to the canvas is the actual size so the canvas must not scale it (i.e. scaling = 1). For now, I use this ugly trick (undefined ratio from parent/shape) to decide to not scale the icon dimensions.